### ISSUE TITLE
rsyslog issue with jboss

### ISSUE THREAD

**Issue by mehrdad2000** [rsyslog/rsyslog#4988]
Hi
after install rsyslog on server and client, and configure jboss have issue on each line, here is the exmple:

client
2022-09-21 09:53:49,763 INFO   [security] traceId:  spanId:  WFLYSEC0002: Activating Security Subsystem

server
2022-09-21T09:53:49.763+04:30 APPname JBoss[18288] WFLYSEC0002: Activating Security Subsystem

this part **[security] traceId:  spanId:**  replaced by **APPname JBoss[18288]**

FYI: also try this:
https://stackoverflow.com/questions/44310596/wildfly-10-sys-log-handler-printing-with-byte-order-mark-feff

got this error:
```
Failed to get the list of the operation properties: "WFLYCTL0030: No resource definition is registered for address [
    ("subsystem" => "logging"),
    ("custom-handler" => "syslog")
]"
```

any idea?

**Comment by davidelang** [comment]
please post your config, and also please write the log out with the template 
RSYSLOG_DebugFormat (please do this on both systems)

that template shows us exactly what rsyslog received and how it parsed it

David Lang

**Comment by mehrdad2000** [comment]
@davidelang added to the post.

**Comment by rgerhards** [comment]
> added to the post.

where?

**Comment by mehrdad2000** [comment]
@radosroka server config, client config

**Comment by rgerhards** [comment]
I was looking for this part: "*also please write the log out with the template RSYSLOG_DebugFormat (please do this on both systems)*"

I suspect that the source subsystem submits an invalidly-formatted message. We will see that once we have the raw message.

**Comment by mehrdad2000** [comment]
@rgerhards added to the post.
FYI: I just add this on jboss domain.xml, and it send jboss log to server directly and not store on client.

```
                <syslog-handler name="SYSLOG">
                    <level name="DEBUG"/>
                    <app-name value="JBoss"/>
                    <server-address value="SERVERIP"/>
                    <port value="514"/>
                </syslog-handler>
```

**Comment by davidelang** [comment]
people have different opinions, but I prefer to send the logs directly to a syslog server. I take the attitude that I want the logs off of the generating server as soon as possible to make them more tamper resistant.

Once you have logs on a central sever, you can then split them up as needed to make them easy to search, and in that case there's less need to have a copy on the original server.

The main argument for keeping a local copy is that it gives you a fallback if there is a network problem or a problem with the central server, that's a trade-off because keeping the logs locally can cause grief on the generating server (not just disk space, but also disk IOPS), I prefer to make my central server redundant and treat logging issues as critical priority.

We still need to see the logs in the RSYSLOG_DebugFormat. As Rainer said, the odds are good that the logs are not being formatted correctly and that's the cause of the problem, but without seeing the raw logs, it's impossible to be sure.

**Comment by mehrdad2000** [comment]
@davidelang but what happen if rsyslog server not available for a while? you will lost logs of that time?

**Comment by davidelang** [comment]
if you are sending over the network with UDP and the server or network is down, 
the logs will be lost.

if you are sending over the network with TCP and the server or network is down, 
logs in flight at the time it went down are lost, and the sender will stall 
(jboss may have soem buffer before the stall)

note that you can send via the network to localhost, that's still a cleaner 
approach than writing to a fale and then reading it in many ways.

David Lang

**Comment by mehrdad2000** [comment]
@davidelang How about RSYSLOG_DebugFormat? did you see that? I added RSYSLOG_DebugFormat in issue.

**Comment by davidelang** [comment]
please also do the debugformat on the client. the decoded message on the server seems to match the data that was sent, so if it's not matching what you expect, then it could be having problems on the client.

Also, it really helps us to post the full config, not just the portion that you think is important, especially when using the legacy config format (as you are), earlier parts of the config (and included files) can affect the behavior of later lines.

**Comment by mehrdad2000** [comment]
@davidelang I add this part on server and client but it doesn't store anything about jboss on client!

```
$template(name=”RSYSLOG_DebugFormat” type=”string” string="Debug line with all properties:\nFROMHOST: '%FROMHOST%', fromhost-ip: '%fromhost-ip%', HOSTNAME: '%HOSTNAME%', PRI: %PRI%,\nsyslogtag '%syslogtag%', programname: '%programname%', APP-NAME: '%APP-NAME%', PROCID: '%PROCID%', MSGID: '%MSGID%',\nTIMESTAMP: '%TIMESTAMP%', STRUCTURED-DATA: '%STRUCTURED-DATA%',\nmsg: '%msg%'\nescaped msg: '%msg:::drop-cc%'\nrawmsg: '%rawmsg%'\n\n")
*.* /var/log/debug.log;RSYSLOG_DebugFormat
```

**Comment by mehrdad2000** [comment]
@davidelang @rgerhards I think it is related to the "formatter pattern" on domain.xml. After I change domain.xml, it's ok
here is the pattern that I use:
<pattern-formatter pattern="%-5p %X{source} [%c{1}] traceId: %X{trace_id} spanId: %X{span_id} %s%E%n"/>

but date format not correct, how can i change date format on server like below?

current time format
2022-09-28T10:38:08+03:30 

expected time format
2022-09-28 10:38:21,096

**Comment by davidelang** [comment]
that's a jboss config thing, I am not familiar with that.

David Lang

**Comment by mehrdad2000** [comment]
@davidelang please forgot about jboss pattern.

here is them main issue on rsyslog server how can I change date pattern?
here is the current config of rsyslog server:

```
$template RemInputLogs, "/var/log/remotelogs/%$YEAR%%$MONTH%%$DAY%/%FROMHOST-IP%/%PROGRAMNAME%.log"
*.* ?RemInputLogs
```

 current time format output
 2022-09-28T10:38:08+03:30

 expected time format
 2022-09-28 10:38:21,096

**Comment by davidelang** [comment]
you mean the format of the timestamp on each line in the file?

to change that you would need to change what tempalte you are using

see https://www.rsyslog.com/doc/v8-stable/configuration/templates.html and 
https://www.rsyslog.com/doc/v8-stable/configuration/properties.html to start

David Lang

**Comment by mehrdad2000** [comment]
@davidelang yes, i mean timestamp.
I saw these pages before and still no idea how it work!
FYI: i edit my last reply.

**Comment by davidelang** [comment]
On Wed, 28 Sep 2022, Mehrdad wrote:

> FYI: i edit my last reply.

thanks for mentioning this, I read these messages via email, so I don't learn 
about edits.

David Lang

**Comment by davidelang** [comment]
on the page https://www.rsyslog.com/doc/v8-stable/configuration/templates.html see the section on reserved template names, that gives the definitions of all the built-in formats.

based on what you are showing, I think you are defaulting to RSYSLOG_FileFormat 

so copy that format (giving it a new name) and change the timestamp definition (either removing it completely, or changing it as explained in the property and property replacer pages), then add it to your action definition the same way that you changed to to the RSYSLOG_DebugFormat for testing.

**Comment by mehrdad2000** [comment]
@davidelang AFAIK structure of rsyslog is like this:
![image](https://user-images.githubusercontent.com/26499665/193401597-e29d0097-0480-4465-9df9-6af83d9714f6.png)

message text is the part that jboss produce it, by first part create by rsyslog server.
as jboss already create log file with timestamp and log level, is it possible to write what exactly jboss create on rsyslog server?

**Comment by davidelang** [comment]
yes, you can create a template that uses $rawmsg in it (you can see what that is 
from the debug logs), I think you will need to add a newline after it.

David Lang

**Comment by mehrdad2000** [comment]
@davidelang 
I have 2 question about "How rsyslog send file?".
Imagine I have 1TB log file on client in this path /var/log/app/mylogfile.log

now I configure it on client then restart service.

1-what happen now? does rsyslog send whole "mylogfile.log"? or it just send logs from now that append to "mylogfile.log"?

2-if it send whole file, how exactly do this? split to smaller size or bucket then send to server? or just send wole file once?

Thanks,

**Comment by rgerhards** [comment]
> 1-what happen now? does rsyslog send whole "mylogfile.log"? or it just send logs from now that append to "mylogfile.log"?

Neither. imfile keeps track of what it already sent. I keeps on from where it stopped the last time.
 
> 2-if it send whole file, how exactly do this? split to smaller size or bucket then send to server? or just send whole file once?

Depends on the config. By default, a single line represents a syslog message and is processed as such. You can switch the behaviour via various config settings.

**Comment by mehrdad2000** [comment]
@rgerhards so whenever service up it send logs from last time service stopped.

1-Does it buffer or store logs in temporary path between service down and up?

2-Is there any way or configuration exist that send remain logs? e.g. Splunk forwarder do the same thing, when service start on client, send whole log file.

3-do you have any link or refrence that explain that part you mention about split file to smaller size?

4-when I enable jboss to logs on file, I have performance issue on my app and need to know if I enable rsyslog to send logs directly to rsyslog server, how much reduce performance impact on my app?

Thanks,

**Comment by davidelang** [comment]
please clarify your question about question #1, does what buffer or store logs? 
rsyslog uses memory buffers internally, and you can configure it to attempt to 
save it to a file on a clean shutdown

read through the imfile documentation, it gives you all the options around the 
options for handling reading from files.

I prefer to have apps send logs to syslog directly rather than having them write 
to a file and then have rsyslog read the file.

David Lang

**Comment by mehrdad2000** [comment]
@davidelang how syslog track file, while service is down? I mean which part send and which part not send and next time service start, start to continue remain log?

**Comment by rgerhards** [comment]
Last state is recorded in side the file-specific state file.
